### PR TITLE
Ignore queue mode for send email to contact action

### PR DESCRIPTION
**Pull request by kuzmany**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/3828
| BC breaks? | 
| Deprecations? | 

This bugfix/feature has been sponsored by @Webmecanik

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
The same workflow like already merged feature https://github.com/mautic/mautic/pull/5414 
Just added option send email to contact immediately in queue mode.

image

#### Steps to test this PR:
Steps to test this PR:
1. Apply PR and got to Configurations > Emails > and Enable queue
2. Create form with Send email to Contact action with new Send immediately option set to No
3: Send form, email should go to queue (verify it in directory app\spool\default)
4. Then Go to form and edit action to Send immediately YES
5. Send from, email should send immediately without queue (again verify it in directory app\spool\default\ - new file shoudn't create)
6. Then change Configurations > Emails > and change Queue to immediately
7. Send form and email should send immediately (again verify it in directory app\spool\default\ - new file shoudn't create)

### CODE REVIEW

File: app/bundles/EmailBundle/Model/SendEmailToContact.php [hunk 1]
```diff
@@ -292,6 +292,11 @@ public function reset()
         $this->mailer->reset();
     }
 
+    public function setSampleMailer()
```
Comment (alanhartless) [inline app/bundles/EmailBundle/Model/SendEmailToContact.php:279]: This will cause all subsequent uses of this class to bypass the default mailer. The original mailer should be stored then restored after the class's use. 

Also please add tests around these new methods.

**Review by alanhartless — CHANGES_REQUESTED** [review]

File: app/bundles/EmailBundle/Model/SendEmailToContact.php (hunk 1 above)
Comment (kuzmany) [reply app/bundles/EmailBundle/Model/SendEmailToContact.php:279]: Work in progress.

Comment (kuzmany) [reply app/bundles/EmailBundle/Model/SendEmailToContact.php:279]: Try  current  solution:
- added $this->temporaryMailer 
- and re-save it before mailHelper reset

Do you have documentation to write tests for Mautic?

**Comment by npracht** [comment]
@justinfortes it has been modified as requested by Alan. Could we add that as candidate for 2.13.0 ?
Thanks !

**Comment by npracht** [comment]
@kuzmany i just tested it again and it works properly.
My concern is that the option is also appearing on the Send email campaign action. Should it be there ? I don't think so.

**Why ?**
1. In campaign you have have high volume of emails sent ----> that's why the queue mode is great. Otherwise do not use the queue mode.
2. In form you won't have a lot of submission at once. Then having it immediately has sense.

You're opinion ?

**Comment by kuzmany** [comment]
@nprach no, shoudn't . I will look on it

**Comment by kuzmany** [comment]
@npracht added new param to form type with_immediately

File: app/bundles/EmailBundle/Model/SendEmailToContact.php (hunk 1 above)
Comment (escopecz) [reply app/bundles/EmailBundle/Model/SendEmailToContact.php:279]: It's not that different than writing PHPUNIT tests for any other PHP app. This model already has some tests:

https://github.com/mautic/mautic/blob/409eb114c3057b71022de5a76cebb43066f4931e/app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php

Just make sure your changes are covered. Here is the command to run the tests: https://github.com/mautic/mautic/tree/409eb114c3057b71022de5a76cebb43066f4931e#phpunit

File: app/bundles/EmailBundle/Form/Type/EmailSendType.php [hunk 2]
```diff
@@ -186,6 +195,30 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 );
             }
         }
+        if (!empty($options['with_immediately'])) {
+            if ($this->coreParametersHelper->getParameter('mailer_spool_type') == 'file') {
```
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/EmailSendType.php]: Lets use `===` to avoid https://twitter.com/jan_linhart/status/786677685988753409

File: app/bundles/EmailBundle/Form/Type/EmailSendType.php [hunk 3]
```diff
@@ -186,6 +195,30 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                'immediately',
+                YesNoButtonGroupType::class,
+                [
+                    'label' => 'mautic.form.action.sendemail.immediately',
+                    'data'  => $default,
+                    'attr'  => [
+                        'tooltip'    => 'mautic.form.action.sendemail.immediately.desc',
+                    ],
+                ]
+            );
```
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/EmailSendType.php:208]: There is something fishy about the indentation here. Shouldn't this line (and some above) move 4 spaces right? How come CS Fixer is good with that?

File: app/bundles/EmailBundle/Helper/FormSubmitHelper.php [hunk 4]
```diff
@@ -63,9 +63,10 @@ public static function sendEmail($tokens, Action $action, MauticFactory $factory
             } elseif (isset($currentLead)) {
                 if (isset($leadFields['email'])) {
                     $options = [
-                        'source'    => ['form', $form->getId()],
-                        'tokens'    => $tokens,
-                        'ignoreDNC' => true,
+                        'source'      => ['form', $form->getId()],
+                        'tokens'      => $tokens,
+                        'ignoreDNC'   => true,
+                        'immediately' => ($factory->getParameter('mailer_spool_type') == 'file' && !empty($properties['immediately'])) ? true : false,
```
Comment (escopecz) [inline app/bundles/EmailBundle/Helper/FormSubmitHelper.php]: Can you include CoreParameterHelper like you did it in EmailSendType? Let's not use the MauticFactory in the new code. And again, use `===`

File: app/bundles/EmailBundle/Model/EmailModel.php [hunk 5]
```diff
@@ -1131,19 +1131,21 @@ public function &getEmailSettings(Email $email, $includeVariants = true)
...
+        $customHeaders        = (isset($options['customHeaders'])) ? $options['customHeaders'] : [];
+        $emailType            = (isset($options['email_type'])) ? $options['email_type'] : '';
+        $isMarketing          = (in_array($emailType, ['marketing']) || !empty($listId));
+        $emailAttempts        = (isset($options['email_attempts'])) ? $options['email_attempts'] : 3;
+        $emailPriority        = (isset($options['email_priority'])) ? $options['email_priority'] : MessageQueue::PRIORITY_NORMAL;
+        $messageQueue         = (isset($options['resend_message_queue'])) ? $options['resend_message_queue'] : null;
+        $returnErrorMessages  = (isset($options['return_errors'])) ? $options['return_errors'] : false;
+        $channel              = (isset($options['channel'])) ? $options['channel'] : null;
+        $dncAsError           = (isset($options['dnc_as_error'])) ? $options['dnc_as_error'] : false;
+        $immediately          = (isset($options['immediately'])) ? $options['immediately'] : false;
```
Comment (escopecz) [inline app/bundles/EmailBundle/Model/EmailModel.php]: I don't see any reason why to move all those line one space to the right. Can you revert it please? It will make code reviews easier if there won't be any unnecessary changes.

**Review by escopecz — CHANGES_REQUESTED** [review]

File: app/bundles/EmailBundle/Helper/FormSubmitHelper.php (hunk 4 above)
Comment (kuzmany) [reply app/bundles/EmailBundle/Helper/FormSubmitHelper.php]: But this is static function. I cannot use $this->coreParametersHelper, can I?

**Comment by kuzmany** [comment]
@escopecz check now, thanks

File: app/bundles/EmailBundle/Config/config.php [hunk 6]
```diff
@@ -612,6 +615,12 @@
                 'arguments' => [],
             ],
         ],
+        'helpers' => [
+            'mautic.helper.form.asubmit' => [
+                'class'     => \Mautic\EmailBundle\Helper\FormSubmitHelper::class,
+                'arguments' => ['mautic.helper.core_parameters'],
+            ],
+        ],
```
Comment (escopecz) [inline app/bundles/EmailBundle/Config/config.php]: Yes, you are right. It is a static method so DI won't help us here. It will need some more refactoring to get rid of MauticFactory in this helper. Please remove this and the constructor in that helper.

Comment (kuzmany) [reply app/bundles/EmailBundle/Config/config.php]: Ok, reverted and ready to re-test

**Review by alanhartless — DISMISSED** [review]

File: app/bundles/EmailBundle/Model/SendEmailToContact.php [hunk 7]
```diff
@@ -289,9 +303,20 @@ public function reset()
         $this->statRepo->clear();
         $this->dncModel->clearEntities();
 
+        if ($this->temporaryMailer != null) {
+            $this->mailer = clone $this->temporaryMailer;
```
Comment (alanhartless) [inline app/bundles/EmailBundle/Model/SendEmailToContact.php]: Should you reset temporaryMailer here to null?

**Review by alanhartless — CHANGES_REQUESTED** [review]

**Comment by kuzmany** [comment]
Label: WIP

**Comment by kuzmany** [comment]
Label: Peending feedback

**Comment by kuzmany** [comment]
Label: Ready to test

**Comment by kuzmany** [comment]
Label: Has Conflicts

**Comment by kuzmany** [comment]
@escopecz Tests added,please review https://github.com/mautic/mautic/pull/5566/files#diff-27b27bab078613ba4d7a049eccb75123 

@npracht can you re-test?

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 8]
```diff
@@ -725,4 +725,109 @@ function () use ($emailMock) {
...
+        $mailer    = new \Swift_Mailer($transport);
+
+        // Mock factory to ensure that queue mode is handled until MailHelper is refactored completely away from MauticFactory
+        $factoryMock = $this->getMockBuilder(MauticFactory::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $factoryMock->method('getParameter')
+            ->willReturnCallback(
+                function ($param) {
+                    switch ($param) {
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: It's better to avoid any logic in the test. To avoid it in this case, use:

```
$factoryMock->expects($this->at(0))->method('getParameter')->with('mailer_spool_type')->willReturn('memory');
$factoryMock->expects($this->at(1))->method('getParameter')->willReturn(''));
```
This code might not work right away. The X `$this->at(X)` means the call index of the class, not the method. It makes it a bit trickier to get the X right.

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 9]
```diff
@@ -725,4 +725,109 @@ function () use ($emailMock) {
...
+                }
+            );
+        $factoryMock->method('getLogger')
+            ->willReturn(
+                new NullLogger()
+            );
+
+        $mockEm = $this->getMockBuilder(EntityManager::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: This 3 lines (and all other mock you create with disabled constructor) can be squeezed into 

`$mockEm = $this->createMock(EntityManager::class);`

**Review by escopecz — CHANGES_REQUESTED** [review]
@kuzmany amazing! I found 2 ways how to make the test simpler.

Could you please test also the `else` part of the new condition? There should be one test for each possible condition in the method.

**Comment by kuzmany** [comment]
@escopecz unit test update.  
What do you mean: "Could you please test also the else part of the new condition?"

File: app/bundles/EmailBundle/Form/Type/EmailSendType.php [hunk 10]
```diff
@@ -186,6 +195,30 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 );
             }
         }
+        if (!empty($options['with_immediately'])) {
+            if ($this->coreParametersHelper->getParameter('mailer_spool_type') === 'file') {
+                $default = (isset($options['data']['immediately'])) ? $options['data']['immediately'] : false;
+                $builder->add(
+                'immediately',
```
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/EmailSendType.php:199]: Move this and the lines that are adding params to the `add` method 4 spaces to the right.

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 11]
```diff
@@ -725,4 +725,102 @@ function () use ($emailMock) {
...
+    public function testSetSampleMailerByTemporaryMailer()
+    {
+        // Use our test token transport limiting to 1 recipient per queue
+        $transport = new BatchTransport(true, 1);
+        $mailer    = new \Swift_Mailer($transport);
+
+        // Mock factory to ensure that queue mode is handled until MailHelper is refactored completely away from MauticFactory
+        $factoryMock = $this->getMockBuilder(MauticFactory::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: These 3 rows can be simplified into `$factoryMock = $this->createMock(MauticFactory::class);`

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 12]
```diff
@@ -725,4 +725,102 @@ function () use ($emailMock) {
...
+                    $tokens['{hash}'] = $event->getIdHash();
+
+                    $event->addTokens($tokens);
+                }
+            );
+        $factoryMock->method('getDispatcher')
+            ->willReturn($mockDispatcher);
+        $routerMock = $this->getMockBuilder(Router::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: These 3 rows can be simplified into `$factoryMock = $this->createMock(Router::class);`

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 13]
```diff
@@ -725,4 +725,102 @@ function () use ($emailMock) {
...
+        $factoryMock->method('getRouter')
+            ->willReturn($routerMock);
+
+        $copyRepoMock = $this->getMockBuilder(CopyRepository::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: These 3 rows can be simplified into `$factoryMock = $this->createMock(CopyRepository::class);`

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 14]
```diff
@@ -725,4 +725,102 @@ function () use ($emailMock) {
...
+        $emailModelMock = $this->getMockBuilder(EmailModel::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: These 3 rows can be simplified into `$factoryMock = $this->createMock(EmailModel::class);`

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 15]
```diff
@@ -725,4 +725,102 @@ function () use ($emailMock) {
...
+            ->setConstructorArgs([$factoryMock, $mailer])
+            ->setMethods(null)
+            ->getMock();
+
+        // Enable queueing
+        $mailHelper->enableQueue();
+
+        $statRepository = $this->getMockBuilder(StatRepository::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: These 3 rows can be simplified into `$factoryMock = $this->createMock(StatRepository::class);`

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 16]
```diff
@@ -725,4 +725,102 @@ function () use ($emailMock) {
...
+                function (Stat $stat) {
+                    $tokens = $stat->getTokens();
+                    $this->assertGreaterThan(1, count($tokens));
+                    $this->assertEquals($stat->getTrackingHash(), $tokens['{hash}']);
+                }
+            );
+
+        $dncModel = $this->getMockBuilder(DoNotContact::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: These 3 rows can be simplified into `$factoryMock = $this->createMock(DoNotContact::class);`

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 17]
```diff
@@ -725,4 +725,102 @@ function () use ($emailMock) {
...
+
+        $translator = $this->getMockBuilder(Translator::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: These 3 rows can be simplified into `$factoryMock = $this->createMock(Translator::class);`

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 18]
```diff
@@ -725,4 +725,102 @@ function () use ($emailMock) {
...
+                new NullLogger()
+            );
+
+        $mockEm = $this->createMock(EntityManager::class);
+
+        $factoryMock->method('getEntityManager')
+            ->willReturn($mockEm);
+
+        $mockDispatcher = $this->getMockBuilder(EventDispatcher::class)
+            ->getMock();
```
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: These 3 rows can be simplified into `$factoryMock = $this->createMock(EventDispatcher::class);`

**Review by escopecz — CHANGES_REQUESTED** [review]
I pointed out all the places where `createMock` can be used. It will make the code shorter and easier to read.

It's OK to create mocks in the test itself if you have only one test in the test class, but when you start adding more it's better to create all the mocks that will be used in other test methods in the `setUp` method. See for example https://github.com/mautic/mautic/blob/staging/app/bundles/WebhookBundle/Tests/Model/WebhookModelTest.php#L32 for inspiration.

File: app/bundles/EmailBundle/Form/Type/EmailSendType.php [hunk 19]
```diff
@@ -186,6 +195,30 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                YesNoButtonGroupType::class,
+                [
+                    'label' => 'mautic.form.action.sendemail.immediately',
+                    'data'  => $default,
+                    'attr'  => [
+                        'tooltip'=> 'mautic.form.action.sendemail.immediately.desc',
+                    ],
+                ]
+            );
+            } else {
```
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: About that comment about testing the `else` part, I meant that you have to write very similar second test so one test would test the `if` part and the other the `else` part. The mock of `$this->coreParametersHelper->getParameter('mailer_spool_type')` will return `file` in one test and something else in the second test.

Comment (kuzmany) [reply app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: Ok, I created testSetSampleMailerByTemporaryMailerSpoolFile  and testSetSampleMailerByTemporaryMailerSpoolMemory
Also I created setUp with mockups, but not time for refactor all class , just my method :(

Comment (escopecz) [reply app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: That's OK. Focus just on your changes. But it shouldn't work the way you did it. That means that the changed code is not tested if you can test both options with one test. I'm looking closely on the test. You should call the `buildForm` method with `FormBuilderInterface` mock as the first argument. Then check what will be added to the builder. And that's the important part. Because if the `mailer_spool_type` === `file` then it will add different params than it it won't be `file`.

Comment (kuzmany) [reply app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: Please help, I've added it but can't move forward

https://github.com/mautic/mautic/pull/5566/files#diff-27b27bab078613ba4d7a049eccb75123R830 
https://github.com/mautic/mautic/pull/5566/files#diff-27b27bab078613ba4d7a049eccb75123R841

Comment (Maxell92) [reply app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: Hello @kuzmany,

There is some misunderstanding I think:
1) We are talking about `EmailSendType` class here, but you linked to test of `SendEmailToContact` class.
2) Just add 2 new tests for this class - first return FALSE and second return TRUE. You will assert that correct call on a builder was performed.

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 20]
```diff
@@ -65,15 +65,44 @@ class SendEmailToContactTest extends \PHPUnit_Framework_TestCase
...
+
+    private $routerMock;
+
+    private $statRepository;
+
+    private $dncModel;
+
+    private $mockEm;
+
+    protected function setUp()
```
Comment (Maxell92) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: Why did you added so many Mocks here? The class `SendEmailToContact` has no dependencies like this. Where is use `MauticFactory` class for example?

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 21]
```diff
@@ -725,4 +755,98 @@ function () use ($emailMock) {
         $errorMessages = $model->getErrors();
         $this->assertCount(1, $errorMessages);
     }
+
+    public function testSetSampleMailerByTemporaryMailerSpoolFile()
+    {
+        $this->testSetSampleMailerByTemporaryMailerSpoolMemory('file');
```
Comment (Maxell92) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: This is interesting, but I think it cannot work:
1) You need different behaviour for each type, right? So you need to write a little different test for each type.
2) Even if the behaviour is same, you can use data providers: https://phpunit.de/manual/6.5/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers

File: app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php [hunk 22]
```diff
@@ -725,4 +755,98 @@ function () use ($emailMock) {
...
+            ->will($this->returnValue(['with_immediately' => true]));
+
+        $model = new SendEmailToContact($mailHelper, $statHelper, $this->dncModel, $translator);
+        $model->setSampleMailer();
+        $this->assertNotNull($model->getTemporaryMailer());
+
+        $model->reset();
+        $this->assertNull($model->getTemporaryMailer());
+
+        $this->builder->expects($this->any())
```
Comment (Maxell92) [inline app/bundles/EmailBundle/Tests/Model/SendEmailToContactTest.php]: The method for `getOptions` is set on line 830. This will take no effect. You need to split this test to 2 separate tests.
Also, please move it to a new class, this test is pretty huge.

File: app/bundles/EmailBundle/Model/SendEmailToContact.php [hunk 23]
```diff
@@ -458,4 +479,12 @@ protected function sendStandardEmail()
         // Now send but don't redispatch the event
         return $this->mailer->queue(true, MailHelper::QUEUE_RETURN_ERRORS);
     }
+
+    /**
+     * @return MailHelper
+     */
+    public function getTemporaryMailer()
```
Comment (Maxell92) [inline app/bundles/EmailBundle/Model/SendEmailToContact.php]: What is this getter for? I did not found any usage of this method.

**Review by Maxell92 — CHANGES_REQUESTED** [review]

File: app/bundles/EmailBundle/Form/Type/EmailSendType.php (hunk 19 above)
Comment (kuzmany) [reply app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: @Maxell92 
 I don't understand. EmailSendType class - there is not TRUE/FALSE logic?  
I am lost now, what test do you need add to PR?

Comment (Maxell92) [reply app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: Hello @kuzmany,
You want to test if / else condition - to be sure correct calls on builder were performed. So there should be 2 tests for the `if (!empty($options['with_immediately'])) {` condition - one for TRUE, second for FALSE.

Comment (kuzmany) [reply app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: @Maxell92 I am building the test

https://github.com/kuzmany/mautic/blob/da8d1634cde20749a9a3d2cbf13916da0e0e5e55/app/bundles/EmailBundle/Tests/Form/Type/EmailSendTypeTest.php 

Can you help how can I provide check If formBuilder generate YesNoButtonGroupType or HiddenType field? 

Thanks

Comment (heathdutton) [reply app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: I'm not sure I'm following this, but is this something we can screenshare and tag-team to get this issue merged?

**Comment by heathdutton** [comment]
Filled out these forms at the same time and behavior was exactly as expected.
screen shot 2018-08-31 at 1 24 59 pm
screen shot 2018-08-31 at 1 25 02 pm
screen shot 2018-08-31 at 1 25 05 pm

File: app/bundles/EmailBundle/Form/Type/EmailSendType.php (hunk 19 above)
Comment (heathdutton) [reply app/bundles/EmailBundle/Form/Type/EmailSendType.php:209]: @kuzmany I think he's suggesting you add one more test after testSendEmailFormBasedOnSpoolType without with_immediately set.

**Comment by kuzmany** [comment]
@heathdutton I think in that time already run cron. But If I tested then works

- 5566 immediate send after form submit
- 5566 not immediate send after cron run (3 min later)

Can you retest?

image

**Comment by kuzmany** [comment]
@heathdutton could you look on it again? my comment above

**Comment by RCheesley** [comment]
@kuzmany this PR has conflicts and also seems to be failing on Travis

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by RCheesley** [comment]
Bumping to 3.2 due to unresolved conflicts

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
Still has outstanding conflicts, I am willing to bump this one more feature release to Mautic 4 and if we have not merged it by then we will close the PR.

**Comment by dennisameling** [comment]
Hi @kuzmany! Thanks for your PR 🚀 we just upgraded Mautic to Symfony 4 and this PR will need to be updated. Since this PR has conflicts now, could you please do a `git merge upstream/features` locally to resolve the conflicts? Thanks in advance!

**Comment by RCheesley** [comment]
@kuzmany will you update this? If so we can consider it for 4.1 but if not let's close this old PR.

**Comment by kuzmany** [comment]
We are gonna switch in M4 to Messenger transport https://github.com/mautic/mautic/pull/10255
Then not need this feature anymore
